### DISCUSSION TITLE
Locode Enhancement Ideas

### DISCUSSION THREAD

**Discussion by mzhukovs** [ServiceStack/Discuss#99] [Ideas]
Love Locode - think it's awesome. Maybe just a couple of features that would make it even more complete, for your consideration:

1. Add declarative attributes to **group props into categories/sections for UI form purposes** - i.e. the new/edit/view pages would display their corresponding inputs as grouped under section headings. Not dissimilar to the [existing Tag attribute](https://docs.servicestack.net/locode/branding#grouping-services-with-tag) for the left menu panel.
- would be nice to also be able to supply a display order, e.g. adding to existing Field/Input/etc. attributes props for display order, which the section would take precedence over (meaning if sections are also used obviously that would just then dictate their order WITHIN that section
  
  
2. **Bulk Upload via Excel** - business users love using Excel, this can be a big time-saver to help them update stuff
- if key is provided, assume update is being performed, o.w. new item creation
- if any errors just spit same Excel file back out with added columns showing all errors - e.g. missing required columns/values, any other validation results by row, if warnings implemented (next pt. separate suggestion) also show warnings and confirm to proceed to ignore them if those are all that are left

3. **FluentValidation - support Warning [Severity level ](https://docs.fluentvalidation.net/en/latest/severity.html)**
- Show any warnings (with corresponding UI warning styling) alongside any errors
- Even if no errors remain but warnings are still present when user attempts to save a new item, add a confirmation pop-up to confirm (proceed with warnings?) or something of the sort

**Comment by mythz** [comment]
Appreciate the feedback, but it would be easier to evaluate, measure interest and discuss feature requests if they were each submitted independently.

For custom forms you can use [field tailwind classes](https://docs.servicestack.net/locode/declarative#custom-fields-and-inputs) to control size and how many inputs on each row using `col-span-*` classes. 

If more customizations are needed you can register [Custom Forms](https://docs.servicestack.net/locode/custom-forms) to control the layout and behavior of different forms.

Finally if you need even more customizations than Locode can provide you can [quickly build Custom Admin UIs](https://www.youtube.com/watch?v=wlRA4_owEsc) as seen in the [Blazor Vue template](https://blazor-vue.web-templates.io/) custom Admin UI created with ServiceStack Vue components in its [admin/index.html](https://github.com/NetCoreTemplates/blazor-vue/blob/main/MyApp/wwwroot/admin/index.html#L122-L136) page.

Bulk Insert/Update via Excel would be a good idea but we wouldn't be adding any dependencies to ServiceStack so it would need to be in .csv format. 

We've not heard of anyone wanting to use FluentValidation Warnings before so we'd need to see some interest in this feature before being able to consider it.

**Reply by mzhukovs** [reply]
Nice, thanks for quick response. Noted RE: submitting ideas separately for future consideration.

On #1 - it's a bit extra work to have to do the Custom Form approach if the only customization you need is to group Fields into sections for presentation in the UI so the ability to do so out-of-the-box would be super convenient, but at least there's a workaround.

On #2 I am sure people would be happy enough with that at least for starters (the .csv format requirement, assuming it would work with Excel's "save as CSV" flavor of it) 

Thanks again.
